### PR TITLE
chore(p2p): dnsaddr recursive resolution

### PR DESCRIPTION
**Pull request by rymnc**
## Linked Issues/PRs

- fixes https://github.com/FuelLabs/fuel-core/issues/2202

## Description

- creates a new module `dnsaddr_resolution` which handles recursive dnsaddr resolution to add to the DHT *without* suffix matching. This way we can connect to all peers behind a domain without specifying the exact `PeerId`, like `/dnsaddr/bootstrap.libp2p.io`.
- asyncifies some of the methods related to mounting the p2p service
- resultifies some of the methods related to mounting the p2p service 

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests
- [x] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [x] I have reviewed the code myself
- [ ] I have created follow-up issues caused by this PR and linked them here

### After merging, notify other teams

[Add or remove entries as needed]

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)
- [ ] [Sway compiler](https://github.com/FuelLabs/sway/)
- [ ] [Platform documentation](https://github.com/FuelLabs/devrel-requests/issues/new?assignees=&labels=new+request&projects=&template=NEW-REQUEST.yml&title=%5BRequest%5D%3A+) (for out-of-organization contributors, the person merging the PR will do this)
- [ ] Someone else?

### CODE REVIEW

File: crates/services/p2p/Cargo.toml
Comment (xgreenx) [inline crates/services/p2p/Cargo.toml:53]: Can we use resolver from `libp2p`?

Comment (rymnc) [reply crates/services/p2p/Cargo.toml:53]: do you mean the same version as libp2p or to use their resolver?

their resolver is in libp2p-dns (https://docs.rs/libp2p-dns/latest/libp2p_dns/tokio/type.Transport.html) and doesn't export any helper functions to resolve dnsaddr's.

Comment (xgreenx) [reply crates/services/p2p/Cargo.toml:53]: I see, they use the same library to do it. Yeah. we need to use the same version to not bloat the `Cargo.lock`

Comment (rymnc) [reply crates/services/p2p/Cargo.toml]: addressed in [4f4b662](https://github.com/FuelLabs/fuel-core/pull/2204/commits/4f4b662c4f9c6e0ac22d381075c58a58723329a5)

File: crates/services/p2p/src/discovery/discovery_config.rs
Comment (xgreenx) [inline crates/services/p2p/src/discovery/discovery_config.rs:194]: What if `dnsaddr` will have another `dnsaddr`, will it work?

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (xgreenx) [inline crates/services/p2p/src/discovery/dnsaddr_resolution.rs:13]: Why do we need trait?

File: crates/services/p2p/src/discovery/discovery_config.rs
Comment (xgreenx) [inline crates/services/p2p/src/discovery/discovery_config.rs:105]: Just curious, have you considered implementation that doesn't require `async` during construction? Maybe it is possible to resolved addresses in the `fn start`?

File: crates/services/p2p/src/discovery/discovery_config.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/discovery_config.rs:194]: yup, resolution is recursive. the test case handles that :)

File: crates/services/p2p/src/discovery.rs
Comment (netrome) [inline crates/services/p2p/src/discovery.rs:274]: I don't see how we can guarantee this future to be immediately ready. What happens if the `DnsResolver::new().await?` is pending for example? As I read this, we'd panic in that scenario. That feels quite brittle to me, in which case it would feel safer to just use the synchronous resolver instead.

My preferred option though would be to allow `build_behavior_fn` to return a function which returns a future, so we can return this future without having to panic if it isn't ready here.

File: crates/services/p2p/src/discovery/discovery_config.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/discovery_config.rs:105]: do you mean here - https://github.com/FuelLabs/fuel-core/blob/1d1560f2fd05b2d28f7df14c98bf65d98fbbcf53/crates/services/p2p/src/p2p_service.rs#L238 ?

File: crates/services/p2p/src/discovery/discovery_config.rs
Comment (netrome) [inline crates/services/p2p/src/discovery/discovery_config.rs:188]: Nit: This reads a bit funky to me. Wouldn't the singular of `multiaddrs` be `multiaddr` rather than `node`? I'd suggest this naming instead:

```suggestion
    let dnsaddr_urls = multiaddrs
        .iter()
        .filter_map(|multiaddr| {
            if let Protocol::Dnsaddr(dnsaddr_url) = multiaddr.iter().next()? {
                Some(dnsaddr_url.clone())
            } else {
                None
            }
        })
        .collect::<Vec<_>>();
```

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (netrome) [inline crates/services/p2p/src/discovery/dnsaddr_resolution.rs:12]: Async traits have been stabilized since [rust 1.75](https://blog.rust-lang.org/2023/12/28/Rust-1.75.0.html), so I'd prefer not to add this declaration for new traits. Instead just have the functions return `impl Future<Output = ...>` and use `async fn` in the implementations.

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (netrome) [inline crates/services/p2p/src/discovery/dnsaddr_resolution.rs:42]: Do we need to do the `Pin<Box<...>>` here? That feels like an artifact of `async_trait` which should preferably be removed (as per my other suggestion), or if you want to keep the `async_trait`, this should happen in the trait implementation and not in this private helper method.

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (netrome) [inline crates/services/p2p/src/discovery/dnsaddr_resolution.rs:99]: Do we assume here that any machine running this test will always be able to look up the "bootstrap.libp2p.io" address in their environment?

For example, what happens if I'm flushing my local DNS cache and disconnect from the internet before running this test? As I read this, the test case would fail in that scenario.

I think we need to introduce a a port to manage the DNS lookup to make this logic testable.

**Review by netrome — CHANGES_REQUESTED** [review]
I hate to block this PR, but there are two pretty substantial issues I can see so far that I would like to have addressed.

1. I don't think we should use `.now_or_never()` to assert that `config.finish().await` is ready in `build_behavior_fn`. It is not clear to me that this future will always be ready when this is invoked.
2. The DNS resolution tests should not be dependent on the host machines local DNS cache or network connection. Instead I'd suggest creating a port for this, which would allow us to create more tests for different DNS configurations.

Moreover, if a port is introduced to allow the DNS lookup logic to be tested I'd be interested in seeing a test case covering recursive lookups to make sure they work as expected.

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (xgreenx) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs:99]: I think introduction of the port will make this tests useless, since it will just test how port itself works. Plus I'm not sure how hard is it I think the introduction of the port will make these tests useless since they will just test how the port itself works. Plus I'm not sure how hard it is to write your own DNS resolver.

I'm okay with the idea that this test fails without a connection or local DNS resolution.

Having a test with real DNS proves that it works in a real environment. The libp2p library tests it in the same way=)

Comment (netrome) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs:99]: While I see your point, I have to respectfully disagree. I think the port will help us make the tests **even more useful**. Right now we can't assert much more than `!multiaddrs.is_empty()`, because there's no way for us to control which DNS records we get when we look up `bootstrap.libp2p.io`. With the port, we for example can test that our logic parses the TXT records correctly.

As for another example, there's currently no way to know from this test if this test is doing a recursive lookup or just a plain single lookup - and this can vary depending on how `libp2p.io` configures their records. With the trait we can write multiple test cases to test and make clear assertions about different scenarios.

I don't think it would be too hard to put a trait between our code and the `TokioAsyncResolver`, because as far as I can see, we're only calling `TokioAsyncResolver::txt_lookup` so we only need to mock one method for the tests.

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs:42]: recursion needs us to pin the future

Comment (netrome) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs:42]: Ah right, makes sense :+1:

File: crates/services/p2p/src/discovery/discovery_config.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/discovery_config.rs]: addressed in 5b4671f1f

File: crates/services/p2p/src/discovery.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery.rs]: deprecated in 5b4671f1f

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs]: removed :)

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (xgreenx) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs]: Okay, I will try to put that in another way=)

The main functionality that I would like to see tested is that we can resolve the real addresses from ipfs(or fuel testnet/mainnet). So, I still want to see the test that uses real `dnsaddr` and verifies that `TokioAsyncResolver` works as expected.

The functionality of `resolve_recursive` and how this function works are not so important to me. 
The approach with the port for internal resolver can help to test the behaviour of the `resolve_recursive` function in different use cases. And it looks like the right call to do it.

It is up to @rymn. I'm okay with doing that in a separate PR just dedicated to better test coverage of the `resolve_recursive` since the main feature is implemented and we have an integration test that covers this feature=)

File: crates/services/p2p/src/discovery/discovery_config.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/discovery_config.rs]: addressed in 0482f8023

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs]: removed the trait

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs]: how about this? I have some dnsaddr's on my personal domain anyway, we can just reuse those with recursion and assert strictly about the multiaddrs this function spits out?

File: crates/services/p2p/src/discovery.rs
Comment (netrome) [reply crates/services/p2p/src/discovery.rs]: Nice, thank you!

**Review by netrome — COMMENTED** [review]
I'll re-review tonight, but my main concerns have been addressed. Thank you!

File: crates/services/p2p/src/discovery/dnsaddr_resolution.rs
Comment (rymnc) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs]: addressed in [8c77c1d](https://github.com/FuelLabs/fuel-core/pull/2204/commits/8c77c1d418b19c39a0d7b25c5d7f5a5b8abc6f7c)

i won't touch the records, i promise 😂

Comment (netrome) [reply crates/services/p2p/src/discovery/dnsaddr_resolution.rs]: Hahah pinky promise :joy: Nice comment also, super helpful thank you! :pray:

File: crates/services/p2p/src/dnsaddr_resolution.rs
Comment (rymnc) [inline crates/services/p2p/src/dnsaddr_resolution.rs:93]: err, this should be _dnsaddr.rymnc.com. patching.

File: crates/services/p2p/src/p2p_service.rs
Comment (xgreenx) [inline crates/services/p2p/src/p2p_service.rs:196]:
```suggestion
    let resolved_multiaddrs = multiaddrs
        .into_iter()
        .filter(|multiaddr| !multiaddr.iter().any(|p| matches!(p, Protocol::Dnsaddr(_))))
        .chain(dnsaddr_multiaddrs.into_iter())
        .collect();
    Ok(resolved_multiaddrs)
```

File: crates/services/p2p/src/dnsaddr_resolution.rs
Comment (xgreenx) [inline crates/services/p2p/src/dnsaddr_resolution.rs:97]: Maybe it is better to use main net dnsaddr=)

**Review by xgreenx — COMMENTED** [review]
LGTM=)

File: crates/services/p2p/src/dnsaddr_resolution.rs
Comment (rymnc) [reply crates/services/p2p/src/dnsaddr_resolution.rs]: addressed in [92f5f22](https://github.com/FuelLabs/fuel-core/pull/2204/commits/92f5f220f77321cc2a5de2b04a20f92c58b4b5f0)

**Review by xgreenx — DISMISSED** [review]

**Review by xgreenx — APPROVED** [review]

File: crates/services/p2p/src/dnsaddr_resolution.rs
Comment (mchristopher) [inline crates/services/p2p/src/dnsaddr_resolution.rs:99]: Can we switch this to `core-test.fuellabs.net`? The records on the `mainnet.fuel.network` record can and will change over time, but I just setup the `core-test` DNS record to be static.

File: crates/services/p2p/src/lib.rs
Comment (netrome) [inline crates/services/p2p/src/lib.rs:8]: Nit: I'd move this declaration to a separate block than the public modules.

**Review by netrome — APPROVED** [review]
Looks good to me :+1: It would be nice to use the test records Mike suggested, but that can be done as a follow-up if you want to get this merged promptly for release.

File: crates/services/p2p/src/dnsaddr_resolution.rs
Comment (netrome) [reply crates/services/p2p/src/dnsaddr_resolution.rs:99]: Thanks for setting this up. I created [this follow-up issue](https://github.com/FuelLabs/fuel-core/issues/2212) to use the `core-test.fuellabs.net` hostname in the test, since this PR got auto-merged after my approval.

**Review by Theghost256 — APPROVED** [review]

**Pull request merged**
